pym/portage/meta_manifest.py

```diff
@@ -0,0 +1,270 @@
+from portage.exception import DigestException, FileNotFound, \
+		InvalidDataType, MissingParameter, PermissionDenied, \
+		PortageException, PortagePackageException
+from portage.const import (MANIFEST1_HASH_FUNCTIONS, MANIFEST2_HASH_DEFAULTS,
+		MANIFEST2_HASH_FUNCTIONS, MANIFEST2_IDENTIFIERS, MANIFEST2_REQUIRED_HASH)
+from portage.localization import _
+
+if sys.hexversion >= 0x3000000:
+	# pylint: disable=W0622
+	_unicode = str
+	basestring = str
+else:
+	_unicode = unicode
+
+
+class MetaManifest(manifest.Manifest):
+	'''Subclass of the Manifest class for Manifest files outside the packages'''
+
+	def sign(self):
+		'''Signs MetaManifest file with the default PORTAGE_GPG_KEY'''
+		filename = self.getFullname()
+		portage_settings = portage.config(clone=portage.settings)
+		portage_portdbapi = portage.portdbapi(portage_settings)
+		gpgcmd = portage_settings.get("PORTAGE_GPG_SIGNING_COMMAND")
+		if not gpgcmd:
+			raise portage.exception.MissingParameter(
+				"PORTAGE_GPG_SIGNING_COMMAND is unset! Is make.globals missing?")
+		if "${PORTAGE_GPG_KEY}" in gpgcmd and "PORTAGE_GPG_KEY" not in portage_settings:
+			raise portage.exception.MissingParameter("PORTAGE_GPG_KEY is unset!")
+		if "${PORTAGE_GPG_DIR}" in gpgcmd:
+			if "${PORTAGE_GPG_DIR}" in gpgcmd:
+				portage_settings['PORTAGE_GPG_DIR'] = os.path.expanduser( \
+					portage_settings.get('PORTAGE_GPG_DIR', '~/.gnupg') \
+				)
+			if not os.access(portage_settings["PORTAGE_GPG_DIR"], os.X_OK):
+				raise portage.exception.InvalidLocation(
+					"Unable to access directory: PORTAGE_GPG_DIR='%s'" %
+					portage_settings["PORTAGE_GPG_DIR"])
+		gpgvars = {"FILE": filename}
+		for setting in ("PORTAGE_GPG_DIR", "PORTAGE_GPG_KEY"):
+			keyvar = portage_settings.get(setting)
+			if keyvar is not None:
+				gpgvars[setting] = keyvar
+		gpgcmd = portage.util.varexpand(gpgcmd, mydict=gpgvars)
+		gpgcmd = portage.util.shlex_split(gpgcmd)
+		gpgcmd = [portage._unicode_encode(arg, encoding=portage._encodings['fs'], errors='strict') for arg in gpgcmd]
+		return_code = subprocess.call(gpgcmd)
+		if return_code == os.EX_OK:
+			os.rename(filename + ".asc", filename)
+			self.signed = True
+		else:
+			raise portage.exception.PortageException("!!! gpg exited with '" + str(return_code) + "' status")
+
+	def find_reporoot(self):
+		'''Finds reporoot'''
+		repodir = self.pkgdir
+		while len(set(['metadata', 'eclass', 'profiles']).intersection(os.listdir(repodir))) == 0:
+			myrepo = repodir.split(os.path.sep)[1:-1]
+			if myrepo[-1] not in myrepo[:-1]:
+				repodir = repodir[:-(len(myrepo[-1])+1)]
+		self.reporoot = repodir
+
+	def find_repolevel(self, path):
+		'''Finds repolevel'''
+		self.find_reporoot()
+		path = path[len(self.reporoot):]
+		myrepo = path.split(os.path.sep)
+		repolevel = len(myrepo)
+		return repolevel
+
+	def find_repotype(self, path):
+		'''Finds the type of the repo directory'''
+		if len(set(['metadata', 'eclass', 'profiles']).intersection(os.listdir(path))) > 0:
```

Comment:
This is faster than listdir:
`if any(os.path.isdir(os.path.join(path, name)) for name in ('metadata', 'eclass', 'profiles')):`